tests/test-lz4-testmode.sh

```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
+
+FPREFIX="tmp-ltm"
+
+set -e
+
+remove () {
+    rm $FPREFIX*
+}
+
+trap remove EXIT
+
+set -x
+
+lz4 -bi0
+datagen > $FPREFIX
+lz4 -f $FPREFIX -c > $FPREFIX.lz4
+lz4 -bdi0 $FPREFIX.lz4 # test benchmark decode-only mode
+lz4 -bdi0 --no-crc $FPREFIX.lz4 # test benchmark decode-only mode
+echo "---- test mode ----"
+! datagen | lz4 -t
+! datagen | lz4 -tf
+echo "---- pass-through mode ----"
+echo "Why hello there " > ${FPREFIX}2.lz4
+! lz4 -f ${FPREFIX}2.lz4 > $FPREFIX-trash
+! datagen | lz4 -dc  > $FPREFIX-trash
+! datagen | lz4 -df > $FPREFIX-trash
+datagen | lz4 -dcf > $FPREFIX-trash
+echo "Hello World !" > ${FPREFIX}1
+lz4 -dcf ${FPREFIX}1
+echo "from underground..." > ${FPREFIX}2
+lz4 -dcfm ${FPREFIX}1 ${FPREFIX}2
+echo "---- non-existing source (must fail cleanly) ----"
+! lz4     file-does-not-exist
```

Comment:
I think there is a problem in the translation of negative tests from `Makefile` to `sh`.

In the `Makefile`, if a test is expected to fail, it can be written, like here `! test_should_fail`.
If, in contrast with expectation, `test_should_fail` actually succeeds,
the `!` will make it fail, and `Makefile` execution will stop immediately.

However, for `sh` script, I don't think it works the same way.
I believe that, if `test_should_fail` succeeds, the `!` will _not_ stop execution.
As a consequence, the next line will be executed,
and the final result of the script will be unaffected by the outcome of `test_should_fail`.

This is obviously not a correct translation of a negative test.

When a test is supposed to fail, there is a need to proceed differently for `sh`. One method is to use the properties of `&&` instead.
The test then becomes something like : 
```
test_should_fail && die "the test should have failed"
```
with `die()` being a `sh` function designed to immediately stop execution of the script with an error code (typically `exit`).

So I'm afraid there's a bit more work to do.

Reply:
I appreciate the explanation. ShellCheck was giving me a similar warning, but I didn't quite understand it. With your explanation, I do get it.

Reply:
Might be worth adding a ShellCheck GitHub Action to the CI. From what I can tell, all these shell scripts would pass, except for one.

Reply:
In a Makefile, the contents of a rule are evaluated ***as a series of shell scripts***, each one run in turn but only if the previous one succeeded. The contents of the rules can be faithfully transcribed into a *.sh file, which @tristan957 was trying to do, and running `set -e`, the "errexit" option, causes the same rules to apply -- namely, that any command which fails causes the script to abort with an error, ***immediately***.

Well, sort of. Actually, `set -e` has some intriguing gotchas. Some things aren't eligible to trigger early "error-exit". One of those things is... `! cmd`. Another is `cmd && othercmd`, as well as `cmd || othercmd`.

Some gotchas of `set -e` are discussed at https://mywiki.wooledge.org/BashFAQ/105

Now, in the case of Make, since each line is actually executed as a *script*, you don't need errexit, because any line that fails will return that error to Make anyway. But for a shell script, you need to force any line that fails to exit explicitly. And that isn't as simple as it initially appears.

So yeah, that's why `! cmd`caused a problem here. Not because there's a difference between a Makefile and a shell script, but because the Makefile is running many small shell scripts, and the actual "real" shell script is just a single shell script.

Reply:
> In a Makefile, the contents of a rule are evaluated as a series of shell scripts

This is good to know

Reply:
> Might be worth adding a ShellCheck GitHub Action to the CI. From what I can tell, all these shell scripts would pass, except for one.

That's a good idea

Reply:
I could add a follow up PR after this one merges, assuming it will.